---
title: "Introduction to Statistical Analysis"
author: "Mark Dunning, Dominique-Laurent Couturier and Sarah Vowler"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document: default
toc: no
---

```{r eval=TRUE, echo=F, results="asis"}
BiocStyle::markdown()
library("knitr")
opts_chunk$set(tidy=FALSE,dev="png",fig.show="as.is",
               fig.width=10,fig.height=4,
               message=FALSE,eval=FALSE,warning=FALSE,echo=FALSE)
```

# Introduction
In this practical, we will use several 'real-life' datasets to demonstrate some of the concepts you have seen in the lectures. We will guide you through how to analyse these datasets in Shiny and the kinds of questions you should be asking yourself when faced with similar data.

To answer the questions in this practical we will be using apps that we have developed using the [Shiny](http://shiny.rstudio.com/gallery/) add-on for the *R* statistical package. **R** is a freely-available open-source software that is popular within academic and commercial communities. The functionality within the software compares favourably with other statistical packages (SAS, SPSS and Stata). The downside is that **R** has a steep learning-curve and requires a basic familiarity with command-line software. To ease the transition we have chosen to present this course using a series of online tools that will allow you to perform statistical analysis without having to worry about learning R. At the same time, the R code required for the analysis will be recorded in the background. You will therefore be able to repeat the analysis at a later date, or pass on to others. As you gain familiarity with R through other courses, you will see how the code generated by Shiny can be adapted to your own needs.

The datasets you will need for this practical should be downloaded and unzipped now:- https://rawgit.com/bioinformatics-core-shared-training/IntroductionToStats/master/CourseData.zip

# Parametric Tests

### 1. The effect of disease on height
A scientist knows that the mean height of females in England is **165cm** and wants to know whether her patients with a certain disease "X" have heights that differ significantly from the population mean - we will use a one-sample t-test to test this. The data are contained in the file **`diseaseX.csv`** and can be analysed online at:-

[http://bioinformatics.cruk.cam.ac.uk/stats/OneSampleTest](http://bioinformatics.cruk.cam.ac.uk/stats/OneSampleTest)

To import the file `diseaseX.csv`; you will need to select the `Choose File` option from the `Data Input` tab and navigate to where the course data are located on your laptop. The right-hand panel of the `Data Input` tab should update to show the Heights of various individuals in the study.

Also, on the `Data Input` tab you will need to change the value of **Hypothesized mean** to the correct value.


------

**Question:** What are your null and alternative hypotheses?

------

```{r}
cat("Null hypothesis: The mean height of female patients with disease X = 165cm 
      (the population mean for females)\n")
cat("Alternative hypothesis: The mean height of female patients with disease X != 165cm 
      (the population mean for females)\n")
```

```{r}
myfile <- "Practical/diseaseX.csv"
sep <- ','
quote <- '"'
header <-  TRUE 
skip <-  0 
data <- read.csv(myfile, header=header, sep=sep, quote=quote,skip=skip)
```



A histogram and boxplot of the `Height` variable will be automatically generated for you. To view it, click on the ***Data Distribution***. You can toggle whether to overlay a density plot on top of the boxplot, or choose different bin sizes for the histogram.

------

**Question:**  Do the data look normally distributed? Based on the plots, is the parametric one-sample t –test appropriate?

------

```{r}
datacol <-  2 
X <- data[,datacol]
colnames(data)[datacol] <- 'X'
library(ggplot2)
ggplot(data, aes(x=X)) + geom_histogram(aes(y=..density..),binwidth=.5,colour='black', fill='white')+ stat_function(fun=dnorm,color='red',args=list(mean=mean(data$X), sd=sd(data$X)))
```

```{r}
cat("In this case the data look normally distributed. Therefore, the one-sample t-test is appropriate.")
```


We are interested in knowing whether the mean height in our sample of patients with disease X is different from that of the general population. Perform a **one-sample t-test** by clicking the ***Statistical Analysis*** tab.

------

**Question:**  What is the mean height in your sample? What is your value of t? What is the p-value? How do you interpret the p-value?

------

```{r}
alternative <- 'two.sided'
mu <-  0 
t.test(X,mu=mu,alternative=alternative)
```

```{r}
cat("Mean height in sample = 171.1cm (95% CI: 169.3-172.9)
t = 7.23, 
df = 19,
p  <= 0.0001.

Under the null hypothesis, the probability of observing a t-statistic as extreme as 7.23, 
is very small P(t <= 7.23 | t >= 7.23) < 0.0001). Therefore, there is strong evidence 
to reject the null hypothesis in favour of the alternative hypothesis. There is strong evidence 
to suggest that the mean height in female patients with disease X is different to the 
population mean height of females of 165cm.")

```


### 2. Biological processes duration

In the file **`bp_times.csv`**, we have the durations of a biological process for two samples of wild-type and knock-out cells (times in seconds). We are interested in seeing whether there is a difference in the durations for the two types of cells – we shall use an **independent t-test** to compare the two cell-types.

These data can be analysed online at [http://bioinformatics.cruk.cam.ac.uk/stats/TwoSampleTest](http://bioinformatics.cruk.cam.ac.uk/stats/TwoSampleTest)

Import the data using ***Choose File*** as before. Make sure that the ***1st column is a factor?*** checkbox is ticked.

------

**Question:**  What are your null and alternative hypotheses?

------

```{r}
cat("Null hypothesis: there is no difference in the duration of the 
biological process for the two cell types.

Alternative hypothesis: there is a difference in the duration of the 
biological process for the two cell types.")
```

Histograms and boxplots to compare the two groups will be created for you automatically. You can also see a basic numerical summary of the data distribution.

------

**Question:**  Do the data look normally distributed for each cell-type? Is the independent t-test appropriate? What statistics are appropriate to report the location (mean or median) and spread (sd or IQR) of the data?

------

```{r}
myfile <- "Practical/bp_times.csv"
sep <- ','
quote <- '"'
header <-  TRUE 
skip <-  0 
data <- read.csv(myfile, header=header, sep=sep, quote=quote,skip=skip)
colnames(data) <- c('variable','value')
p <- ggplot(data, aes(x=value)) + geom_histogram(aes(y=..density..),colour='black',fill='white') + facet_wrap(~variable) + stat_function(fun=dnorm,color='red',args=list(mean=mean(data$value), sd=sd(data$value)))
p
```

```{r}
cat("The data do appear to be approximately normally distributed as we could 
easily draw a bell shape over each of the two histograms. The independent t-test is appropriate")
```

In order to apply the correct statistical test, we need to test to see if the variances of the two groups are comparable. This is tested for us automatically in the Shiny app. Click the ***Statistical Analysis*** tab to see the result of the "F-test". However, it is often easier to eye-ball the data to assess the variances.

------

**Question:** What do you conclude from the p-value of this test. Does this agree with your impression of the variances from the boxplot and histograms? How does it influence what test to use?

------

```{r}
cat("The test yields a p-value of 0.03568, which is sufficient evidence 
    to reject the null hypothesis that the variances of the two groups are the same. 
    Therefore we should apply Welch's correction.")
```


Now use the appropriate two-sample t-test to compare the durations of the two groups.

------

**Question:** What is your value of the test statistic? What is the p-value? How do you interpret the p-value? 

------
```{r}
alternative <- 'two.sided'
paired <- FALSE
var.equal <-FALSE
t.test(value~variable,data=data,alternative=alternative,paired=paired,var.equal=var.equal)

```

```{r}
cat("When we run the independent (unpaired) t test, a formal comparison of the variances 
between the two groups is automatically run. The corresponding p-value is 0.00125 which indicates 
evidence to reject the null hypothesis.")
```



### 3. Blood vessel formation

In blood plasma cancer, there is an increase in blood vessel formation in the bone marrow. A stem cell transplant can be used as a treatment for blood plasma cancer. The bone marrow micro vessel density was measured before and after treatment for 7 patients with blood plasma cancer.

We are interested in seeing whether there is a decrease in the bone marrow micro vessel density after treatment with a stem cell transplant. We will use a paired two-sample t-test to compare the before and after bone marrow micro vessel densities. 

These data can be analysed online at [http://bioinformatics.cruk.cam.ac.uk/stats/TwoSampleTest](http://bioinformatics.cruk.cam.ac.uk/stats/TwoSampleTest)

The data are contained in the file **`bloodplasmacancer2.csv`**. Import the data, making sure that ***`1st column is a factor`*** is *not* ticked. Now choose whether you will be performing a paired test or not by ticking the **Paired Samples?** box under ***Are your samples paired?***. 

------

**Question:** What are your null and alternative hypotheses?

------

```{r}
cat("Null hypothesis: the bone marrow micro vessel density after treatment is greater 
than or equal to the bone marrow micro vessel density before treatment. 

Alternative hypothesis: the bone marrow micro vessel density after treatment 
is less than the bone marrow micro vessel density before treatment.")

```

View the histogram and boxplot of the paired differences on the ***Differences*** tab. 

------

**Question:** Do the differences look normally distributed? Is the paired t–test appropriate?

------


```{r results='hide'}
myfile <- "Practical/bloodplasmacancer2.csv"
sep <- ','
quote <- '"'
header <-  TRUE 
skip <-  0 
data <- read.csv(myfile, header=header, sep=sep, quote=quote,skip=skip)
data <- reshape2::melt(data)
colnames(data) <- c('variable','value')
library(ggplot2)
p <- ggplot(data, aes(x=value)) + geom_histogram(aes(y=..density..),colour='black',fill='white') + facet_wrap(~variable) + stat_function(fun=dnorm,color='red',args=list(mean=mean(data$value), sd=sd(data$value)))
newDf <- do.call(cbind,split(data$value,data$variable))
Diff <- data.frame(Difference=newDf[,1] - newDf[,2])
p2 <- ggplot(Diff,aes(x=Difference)) + geom_histogram(aes(y=..density..),colour='black',fill='white') + stat_function(fun=dnorm, color='red', args=list(mean=mean(Diff$Difference), sd=sd(Diff$Difference)))
p <- gridExtra::grid.arrange(p,p2)
p

```



```{r}
cat("From this histogram it is difficult to tell whether the differences between 
  the densities before and after treatment are normally distributed. In situations 
  like this, we may need to draw on the experience of similar sets of measurements. ")
```


We are interested in seeing whether there is a decrease in the bone marrow micro vessel density after treatment with a stem cell transplant. 

------

***Question:*** Is this a one-tailed or two-tailed test?

------

```{r}
cat("One-tailed as we are only interested in a decrease. 
Usually a two-sided test is preferred unless there is a strong argument 
for a one-sided test. In this case our treatment is only considered to be effective 
if we see a reduction in the bone marrow micro vessel density after treatment. 
Observing an increase in bone marrow density after treatment would lead to the same 
action/conclusion as if no difference had been observed – the treatment might be 
dropped from the research programme (but bear in mind here, 
the sample size is small and so only large differences may be detected).")
```

Now select the correct options in the ***Statistical Analysis*** tab in order to perform the analysis. Ensure you select the one- or two-tailed test as appropriate. 

------

**Question:** What is the mean difference? What is your value of t? What is the p-value? How do you interpret the p-value? Why does ticking / unticking the equal variances have no effect?

------

```{r}
alternative <- 'greater'
paired <- TRUE
var.equal <-FALSE
t.test(value~variable,data=data,alternative=alternative,paired=paired,var.equal=var.equal)

```

```{r}
cat("Under the null hypothesis, the probability of observing a t-statistic as extreme as 1.84, is 0.06, 
slightly greater than 0.05, our nominal significance level. Our result is borderline. 
There is insufficient evidence to reject the null hypothesis. 
Therefore, we might conclude that there is an association of a decrease in bone marrow micro vessel 
density after treatment with bone marrow transplant. It is important to note the small sample size here. 
Studying just 7 patients means we will only be able to detect large differences.")
```


### 4. Gene Expression in Breast Cancer patients

A gene expression study was performed on patients categorised into positive and negative Estrogen Receptor (ER) groups. It is well-known that ER positive patients have more treatment options available and thus have more better prognosis.

The gene NIBP was measured as part of this study and the results are available in the file `NIBP.expression.csv`. We are interested to see if the expression level of the gene is different between ER positive and negative patients.


------

**Question:** What are your null and alternative hypotheses?

------

Now conduct an independent two-sample t-test to see if there is a difference in expression between the two groups.

------

**Question:** What is the p-value from the test? Do we achieve statistical significance at the 0.05 level?

------

Look closely at data distribution, calculated means for each group and the estimated confidence interval

------

**Question:** Is the finding likely to hold Biological significance? Would you be willing to put further resources into validating the finding?

------
